fair/2.0.0/emissions_driven.py

```diff
@@ -0,0 +1,186 @@
+        alpha[...,t] = unifiedtools.calculate_alpha( G=G,\
+                                                         G_A=G_A,\
+                                                         T=T[t-1],\
+                                                         r0=r0,\
+                                                         rC=rC,\
+                                                         rT=rT,\
+                                                         rA=rA,\
+                                                         g0=g0,\
+                                                         g1=g1)
+        C[...,t], R, G_A = unifiedtools.step_concentration( emissions = inp_ar[np.newaxis,...,t],\
+                                                                a = a,\
+                                                                dt = timestep[t],\
+                                                                alpha = alpha[np.newaxis,...,t],\
+                                                                tau = tau,\
+                                                                R_old = R,\
+                                                                PI_conc = PI_conc,\
+                                                                emis2conc = emis2conc)
+        RF[...,t] = unifiedtools.step_forcing( C=C[...,t],\
+                                                   PI_conc=PI_conc,\
+                                                   f1=f1,\
+                                                   f2=f2,\
+                                                   f3=f3)
+        S,T[t] = unifiedtools.step_temperature( S_old=S,\
+                                                    F=np.sum(RF[...,t],axis=0) + ext_forcing[t],\
+                                                    q=q,\
+                                                    d=d,\
+                                                    dt=timestep[t])
+        G += inp_ar[t]
+    res = {'C':C, 'RF' : RF, 'T' : T, 'alpha':alpha}
+    return res
+
+def _convert_df_to_numpy(inp_df):
+    """
+    Convert input df to numpy array with correct order for running
+
+    Parameters
+    ----------
+
+    inp_df : :obj:`pd.DataFrame`
+        Input :obj:`pd.DataFrame` to be converted to a numpy array. Column index is strictly `pd.MultiIndex` with two levels
+
+    Returns
+    -------
+    :obj:'np.ndarray'
+        df converted to a numpy array with correct order for running. The numpy array is ordered as follows: [Level 1, Level 2, Index]
+        Where Level 1 & Level 2 are the levels of the columns ordered alphabetically by level label
+        Each column is then ordered using the ASCII sort order
+
+    """
+    # here you use the metadata from the dataframe to make sure you get the
+    # column order right
+
+    #sorts the df so ordering is 'correct' within levels/index
+    df = inp_df.sort_index(axis=0).sort_index(axis=0)
+    df.columns.names
+    n = len(df.columns.names)
+    while True:
```

Comment:
is this sorting the columns? If so, I'm pretty sure this can be done more simply https://stackoverflow.com/a/11067072. If not, can you split this into a separate function and give it a name which describes what it's doing please